### ISSUE TITLE
CPU spikes when loading large file

### ISSUE THREAD

**Issue by mwanchap** [JimBobSquarePants/UmbracoFileSystemProviders.Azure#112]
Firstly, thanks for all your hard work on this project - it's been a huge help for us!

We manage a bunch of Umbraco-based sites and we recently had a situation where a large-but-not-ridiculous number of requests for a large-ish PDF (~10mb) was causing the Azure app service plan hosting the sites to have serious CPU usage spikes, > 90% for a period of several hours as long as the requests continued.  These were all direct file requests to a /media/... url, so I don't think much other Umbraco-related work was happening aside from retrieving the file.  This seems to only occur on large files as load testing on the site hasn't shown the same behaviour when requesting a large number of smaller files (e.g. images on a page).  I was wondering what your thoughts were on the potential cause and any ideas on mitigations (aside from just putting the file elsewhere or using a CDN).

**Comment by JimBobSquarePants** [comment]
Hi @mwanchap 

We're not doing anything intensive at all with our codebase. We merely call the underlying Azure API's to retrieve the content over the wire.

**Comment by SteveVaneeckhout** [comment]
I had the same problem last year. Large number of requests to PDF files from the same IP addresses for hours and sometimes days. Based on my research it's an older version of Adobe Reader that is to blame. Adobe reader doesn't understand the response from IIS and keeps rerequesting the same file until the user closes the tab/browser.

If you still have the issue, you could call the MediaFilesystemProvider and return the URL to the blob storage. That way the requests bypass your webserver completely.

**Comment by mwanchap** [comment]
Yeah that's similar to what we ended up doing, although in our case
Umbraco's blob storage account was private, so we moved the file to a
public one and linked via the URL as you've suggested. Thanks for the
research about Adobe reader, how did you figure that out? My guess was
email server security scanners, since we have lots of PDFs in Umbraco like
this and have never encountered this behavior, until we emailed links to
one around to all our customers.

**Comment by SteveVaneeckhout** [comment]
I think I found it by Googling IIS PDF. In the end I found the following links:

- [You cannot open some IIS 7.5-hosted PDF documents by using a Web browser that has the Adobe PDF Reader plug-in enabled](https://support.microsoft.com/en-us/help/979543/you-cannot-open-some-iis-7-5-hosted-pdf-documents-by-using-a-web-brows)
- [Handling of Byte-Range requests in IIS 7.5 Web Server causes Adobe Reader web browser plug-in to fail loading PDF documents](https://helpx.adobe.com/archive/acrobat/acrobat-9-troubleshooting.pdf) (page 244)
